### ISSUE TITLE
SIGHASH_TOKEN

### ISSUE THREAD

**Issue by bitjson** [bitjson/cashtokens#11]
Just opening an issue to track a change [discussed here](https://bitcoincashresearch.org/t/chip-2022-02-cashtokens-token-primitives-for-bitcoin-cash/725/22?u=bitjson):

> This is a great point, thanks for noticing [@im_uname](https://bitcoincashresearch.org/u/im_uname)! As [@bitcoincashautist](https://bitcoincashresearch.org/u/bitcoincashautist) mentioned, requiring `OP_RETURN` for token destruction makes cleaning up “token dust” much more expensive than some sort of `SIGHASH_TOKEN`. A signing serialization type flag also simplifies away the awkward destruction policy difference between fungible or immutable tokens and minting or mutable tokens (where the former currently cannot be implicitly destroyed, but the later can).
>
> In fact, to prevent offline signers from being mislead by omitting token information in signing requests, we also need to add the token information directly to the signing serialization. Maybe we include the full contents of the token output prefix (same encoding) [after `value` in the algorithm](https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/replay-protected-sighash.md#digest-algorithm) for `SIGHASH_TOKEN` signatures?
>
> If you’d be interested in sending a PR, I’d definitely appreciate it! (I’m also happy to write the update or any relevant rationale, just let me know what you’re interested in doing.)

**Comment by bitjson** [comment]
@imaginaryusername is already working on this PR, but just so we don't forget: 

`SIGHASH_TOKEN` should also directly commit to contents of all other UTXOs (if `SIGHASH_ALL`) to ensure offline signers can't be [fooled into destroying tokens or burning funds via fees, especially in CashFusion use cases](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-April/017801.html). (Note that BTC's Taproot upgrade recently deployed [a completely new signing serialization algorithm](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-16) that improves this situation for Taproot spends on BTC; we can simply augment our existing algorithm.)

**Comment by imaginaryusername** [comment]
PR is in: https://github.com/bitjson/cashtokens/pull/14

About committing to contents of all other UTXOs: See the PR, intention is to keep current behavior of flags, so `SIGHASH_ALL` will already include such behavior.

**Comment by bitjson** [comment]
@cculianu was reviewing this and suggested even this `SIGHASH_TOKEN` approach isn't necessary to protect users from unintentionally burning tokens; outdated clients will not understand the token encoding, so you'll need more than one software bug to line up before users are impacted. 

I think I agree, we're probably being overly cautious on this front because it was such and issue for the SLP ecosystem.

I [previously reviewed how many major wallets work internally](https://bitcoincashresearch.org/t/chip-2021-01-pmv3-version-3-transaction-format/265/48#most-wallets-_dont_-require-updates-for-pmv3-10), and almost all of them pre-parse UTXOs in the wallet backend before sending a formatted message to the wallet software. **This configuration happens to make them quite a bit more susceptible to token-burning bugs**: if the backend parses the token-containing UTXO and sends it to the client without token information (or with token information, but the client can ignore the additional information), wallets could inadvertently burn those tokens. 

**It would be reasonable to consider this a bug in the wallet backend** – backends shouldn't strip token information or otherwise assist outdated wallets in burning tokens. So at the very least, if we drop explicit burning and don't replace it with `SIGHASH_TOKEN`, we should mention this risk explicitly in the spec (to ensure wallet backend implementors are aware of the potential bug).

Of course, that more hands-off approach still leaves the possibility that some wallet backends will implement the incorrect behavior (either by incompetence or maliciousness), leading some users to accidentally burn newly received tokens. It would be nice if we could find ways to help prevent those cases.

One significant downside of the simple `SIGHASH_TOKEN` approach (and the reason I've been slow to merge #14) is that we'd be spending another sighash bit on a one-time activation – I'd really prefer we avoid leaving such "activation baggage" in the protocol. One option that wouldn't leave a footprint in the protocol: we could implement a `SIGHASH_SCRIPTPUBKEYS` that solves the more general issue of [wallets being tricked into burning either BCH or tokens](https://github.com/bitjson/cashtokens/issues/11#issuecomment-1085000342), and we could **make `SIGHASH_SCRIPTPUBKEYS` required by token transactions for standardness for a year or so**. Then users are somewhat protected from unintentional bugs – e.g. if a wallet backend pre-parses a token UTXO and manages to send it to an unupgraded wallet (and that wallet ignores the new `token` field), that wallet will still have trouble unintentionally burning the funds. 

By burning tokens during this period (e.g. between 2023 and 2024) without `SIGHASH_SCRIPTPUBKEYS`, the transaction is non-standard, and will likely be rejected by whatever node the wallet uses to push transactions (or hopefully at least fail to propagate/confirm, if the wallet backend + node are maliciously trying to get a non-upgraded user to burn newly received tokens). After e.g. one year, we remove the `SIGHASH_SCRIPTPUBKEYS` standardness requirement, and the protocol retains no activation baggage.

**Comment by cculianu** [comment]
Thanks @bitjson for summarizing out conversation!

> **This configuration happens to make them quite a bit more susceptible to token-burning bugs: i**
> ...
> wallets could inadvertently burn those tokens.

Ouch. That would be bad.  Well then if that's the case, we can totally do `SIGHASH_TOKEN`. 

Just to be clear I was playing devil's advocate with my initial resistance to the `SIGHASH_TOKEN` bit. However, if you did some cursory evaluation and think it's warranted, I trust your judgement on this one.  We can totally do it. Perhaps an ounce of prevention is worth a pound of cure in this case. 

> So at the very least, if we drop explicit burning and don't replace it with SIGHASH_TOKEN, we should mention this risk explicitly in the spec (to ensure wallet backend implementors are aware of the potential bug).

Sure, we can do either/or.  I'm not opposed to either the more I think about it. 

> is that we'd be spending another sighash bit on a one-time activation  

To be frank that was what bothered me about it initially too... and hence my trying to straw-man argue against it -- to test if resolve was strong that it's really justified.  I actually don't even know *what* the real risk is of inadvertent token burn anyway so... it's hard for me to evaluate if this cost is justified or not...

> After e.g. one year, we remove the SIGHASH_SCRIPTPUBKEYS standardness requirement, and the protocol retains no activation baggage.

Hmm! This is a clever idea.  So -- what about the consideration that we are "spending" a bit on this?  Would this bit have some other potential meaning in the future someday if we run out of bits perhaps long after it has expired?

**Comment by imaginaryusername** [comment]
If a sighash flag is initially required for standardness but later not, it would mean another mass change in software is needed - and if we don't enforce that change, software will be patchwork "changed" meaning we lose that bit in practice anyway. Going standardness to later remove does not sound like a good idea to me, we should either not do the bit (if we decide that backends are unlikely enough to do the offending behavior), or just do the bit permanently.

**Comment by bitjson** [comment]
I should have clarified – I think `SIGHASH_SCRIPTPUBKEYS` is critical for security of token-aware wallets in general. The much older [issue of offline wallets being tricked into burning funds as fees](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-April/017801.html) is even easier to exploit for a profit (even without the help of a miner) against a lot of covenant+token designs, and it would likely be an issue even for non-hardware wallets. (I just forgot to include it in the first draft of the spec.) So `SIGHASH_SCRIPTPUBKEYS` is an important security feature for all token-supporting wallets, and it's also useful in preventing that fee-buring attack on offline wallets (the reason it was added to BTC's Taproot upgrade). It just happens to also offer a solution to this issue too.

Anyways, I think I agree with @imaginaryusername in that trying to use standardness makes things messier. 

Yesterday I was thinking that using standardness is the more conservative approach – it would offer most of the security benefits (prevent both token burning attacks and the kind of bug described about with wallet backends) and it wouldn't necessarily prevent esoteric use cases where a token transaction needs to have signatures without `SIGHASH_SCRIPTPUBKEYS`.

However, I'm now thinking that requiring `SIGHASH_SCRIPTPUBKEYS` for token transactions by consensus is actually more conservative – we would have a single consistent behavior in the immediate term, and 1) its possible no one will ever want to create token transactions without `SIGHASH_SCRIPTPUBKEYS`, and 2) if they do, a future consensus modifying upgrade is more reliable than a standardness upgrade (some node implementations may fail to relax a standardness requirement, but they would be forked off the chain if they failed to relax a consensus requirement). So consensus enforcement is better in both cases. 

So I'm thinking the best solution for both problems is to require `SIGHASH_SCRIPTPUBKEYS` by consensus for any transactions involving tokens, and we don't need a separate `SIGHASH_TOKEN`. 

(I'll working on that in my next PR.)

**Comment by A60AB5450353F40E** [comment]
`SIGHASH_SCRIPTPUBKEYS` means you'd have both the prevout script + the modified input script? Meaning, in case of p2sh, you'd have the (PFX+{P2SH hashlock})+redeem script?

**Comment by cculianu** [comment]
Just to leave a note here -- if someone wants to go with `SIGHASH_SCRIPTPUBKEYS` and make the signature commit to a hash of *all* the `scriptPubKeys` -- one should *very clearly* specify in the spec for that facility that it is the `scriptPubKey` blob as it **appears in the serialized txn** (and not the "logical `scriptPubKey`" as is separated out into memory after deserializing the token data at the front of it).  I know that's obviously the intent just want the specs to be clear on that ...

**Comment by cculianu** [comment]
As an aside, and along the lines of the above -- another thought: We could not modify the signature hash type bits at all. A different hack we could do (and which would defend tokens!) is just enforce that the "scriptCode" or scriptpubKey that is being hashed in the preimage be the **full serialized blob including token data**.  This wouldn't be quite as secure as the `SIGHASH_SCRIPTPUBKEYS` approach but would **most definitely** prevent burning from unupgraded wallets that are not token aware in 99.999% of cases I think.

**Comment by imaginaryusername** [comment]
On top of that, we can't just go out there and force a signature coverage change on _all_ wallets, that'll be network effect suicide.

**Comment by cculianu** [comment]
What I proposed has very little impact on most wallets.  

It wouldn't be that bad to change behavior of OP_CODESEPARATOR.  It's rarely used.

**Comment by cculianu** [comment]
You know a problem with this is that it's mixing layers in a way. It's not clear *where* the sighash byte lives in the txn until you EVALUATE the txn in the vm.  You you would have to enforce this consensus rule only when you hit an `OP_CHECKSIG*` op-code.  Something about this feels extremely hacky and fragile.  No other consensus rule about what is a valid txn and what is not is applied in this way.  It's completely outside the norm.  Not sure if that means implementation would be tricky or not.

It just bothers me .. is there any way you guys can just be ok with burns?  Can we get some numbers on which wallets out there would be candidates for accidental burns if unupgraded?

Electron Cash would not be.  It would never be able to sign these txns that have tokens in them, due to the way it works it would never recognize thjs inputs it is signing as "mine" because unless it is upgraded it would go back and see the scriptPubKeys of the inputs and think they are nonsense and not "mine" so it could never match them to a private key with which to sign.

What other wallets do people use? Bitcoin.Com? Neutrino? Automated systems in backends?

**Comment by A60AB5450353F40E** [comment]
Had a little brainstorming session with Calin, conclusion is this:

1. There's a problem with adding a new bit: forcing it to be used for signing token TXs with the bit requires "mixing layers" - the signing code currently doesn't care about what the signed data means, it just collects the data from the TX and signs it - it shouldn't start caring now about what the data means.

However, we want to make token TXs fully incompatible with old software to prevent burns, so we need to do it WITHOUT a bit - but change the signagure format in some way when it builds the preimage with token data in it, we suggest:
- include the whole PFX annotation of inputs, BUT -  prepend it with byte 0x01 (which could later be incremented for new PFXes). This will be unexpected by old sofware which would omit it because it's be making the PFX part of "locking script".

**Comment by cculianu** [comment]
Yes we concluded the best of all worlds (which doesn’t require something as vast  SIGHASH_SPKs) would be to modify the definition of “modified locking script” in the token case only to have the modified locking script commit to the token data as well in some way!

if you do it in such a way so that Unuograded software cannot just copy-paste data it saw to sign… it would fix the issue for us 100%

I believe bchautist volunteered to spec this out 

the gist is you redefine modified locking script to be: 0x1 serialized-token-data-with-pfx-byte + scriptCode

**Comment by cculianu** [comment]
Ok I went ahead and implemented this. Note this is what it looks like basically:

```c++
        CHashWriter ss(SER_GETHASH, 0);
        ss << txTo.nVersion << hashPrevouts << hashSequence  << txTo.vin[nIn].prevout; // this was here before
        if (prevTxOut.tokenDataPtr) {
            // New! For tokens. If we had tokenData we inject it as [0x1 token::PREFIX_BYTE ID ..]
            // (just the serialized token data prefix, sans scriptPubKey), right *before* scriptCode.
            ss << uint8_t{0x1} << token::PREFIX_BYTE << *prevTxOut.tokenDataPtr;
        }
        // this bit is same as before (even if no token this is what came next normally)
        ss << scriptCode << prevTxOut.nValue << txTo.vin[nIn].nSequence << hashOutputs << txTo.nLockTime;
        // Sighash type
        ss << sigHashType;

        return ss.GetHash();
```

The *new* code is that `if (prevTxOut.tokenDataPtr) {` branch which is only going to be taken after activation when tokens are enabled.  

So basically to spend a token, you need to serialize the preimage a bit differently.  This will 100% fix burning on unupgraded software because unupgraded software would have serialized **either** `ss << PREFIX_BYTE << *prevTxOut.tokenDataPtr << scriptCode << prevTxOut.nValue`  **or** `ss << scriptCode << prevTxOut.nValue` in that section.. **depending on the nature of the defect of the unupgraded wallet**.   In **no case** will it ever generate the hash preimage using the extra `0x1` we are injecting into the stream...

**Comment by A60AB5450353F40E** [comment]
Yeah! Also note, talking with imaginary yesterday I realized we don't even need the magic byte if you put the blob before the `scriptCode`'s length (and don't count the PFX for the length).

**Comment by cculianu** [comment]
> Yeah! Also note, talking with imaginary yesterday I realized we don't even need the magic byte if you put the blob before the `scriptCode`'s length (and don't count the PFX for the length).
> 
> code doesn't seem to add the length, so is the length encoded "inside" the `scriptCode` or are the [docs](https://documentation.cash/protocol/blockchain/transaction/transaction-signing#preimage-format) wrong?

This is true! I got rid of that `0x1` byte from my code in the latest commit.

**Comment by A60AB5450353F40E** [comment]
Ok, I implemented it for V8, I'll leave it to Jason to update CT2.0.
We can do the same for outputs since they're logically separate things - and here we WANT to break the signature format when it involves tokens so it's all good, how's this: https://gitlab.com/0353F40E/group-tokenization/-/commit/0ddbfd243eab27749e668218a27bbfc446b4fe96

**Comment by cculianu** [comment]
I'm having trouble understanding this since it's morning and I'm burnt out.  The idea is the outputs for hashOut are going to be the same serialization format as you would encounter in a TXN.  That is, nothing special needs to be done here.  No magical breaking up for anything.  If that's what your spec is saying, great.  I think that's what it's saying.

**Comment by A60AB5450353F40E** [comment]
It was not, following the same logic I thought that maybe it'd be cleaner to put it in its own box... but chatting with you now I understand it's simpler to keep it as it is: add the whole output blob as it is (pfx would be automatically included). Updated: https://gitlab.com/0353F40E/group-tokenization/-/commit/6094119dd5c557b5b4132f415280448446e5aa96

**Comment by cculianu** [comment]
Thank you man I appreciate it 👍

**Comment by bitjson** [comment]
Thanks everyone for the discussion here! Thanks @cculianu for realizing we don't need to reserve a `SIGHASH` bit just for tokens; simply including tokens in the signing serialization is a much more natural solution. 🤦‍♂️

Anyways, I think this is resolved in `v2.1.0`. I'll close this issue for now, but let me know if I should re-open.

**Comment by bitjson** [comment]
This is not a concern now that implicit burns are supported, but leaving a thought here that would waste space in the CHIP rationale (there's not really even an argument for "explicit burns" anymore): probably the biggest reason not to require some explicit burning of tokens in outputs is that it would open up a lot of covenants to denial of service attacks. If the covenant doesn't leave a code path open to allow burning of "token spam", the covenant could be frozen by an unexpectedly received token.
